File: comp-arch-2/src/two_dimensional_array.cpp
//------------------------------------------------------------------------------
// two_dimensional_array.cpp - содержит функции обработки двумерного массива
//------------------------------------------------------------------------------

#include "two_dimensional_array.h"

//--------------------------------------------
// Ввод элементов двумерного массива из потока
void TDArray::In(std::ifstream &ifst) {
    ifst >> this->dimensionality;
    int size = dimensionality;
    int col = 0;
    int row = 0;
    for (int i = 0; i < size * size; ++i) {
        col = i % size;
        row = i / size;

        ifst >> elements[row][col];

        ++col;
    }
}

//------------------------------------------------------
// Случайный ввод элементов двумерного массива из потока
void TDArray::InRnd() {
    dimensionality = Random();
    int size = dimensionality;
    int col = 0;
    int row = 0;
    for (int i = 0; i < size * size; ++i) {
        col = i % size;
        row = i / size;

        elements[row][col] = Random();

        ++col;
    }
}

//---------------------------------------------
// Вывод элементов двумерного массива из потока
void TDArray::Out(std::ofstream &ofst) {
    int size = dimensionality;
    int col = 0;
    int row = 0;
    ofst << "Two dimensional array:\n"
         << "Dimensionality: " << dimensionality << "\n"
         << "Arithmetic mean: " << ArithmeticMean() << "\n"
         << "Elements: \n";
    for (int i = 0; i < size * size; ++i) {
        col = i % size;
        row = i / size;

        ofst << elements[row][col] << " ";

        ++col;

        if (col == size) {
            ofst << "\n";
        }
    }
}

//--------------------------------------------
// Вычисление среднего арифметического массива
double TDArray::ArithmeticMean() {
    int size = dimensionality;
    int count = size * size;
    double sum = 0;
    int col = 0;
    int row = 0;

    for (int i = 0; i < size * size; ++i) {
        col = i % size;
        row = i / size;

        sum += elements[row][col];

        ++col;
    }

    return sum / count;
}

File: comp-arch-3/src/two_dimensional_array.py
from matrix import Matrix
from random_generator import get_rand_int

from typing import IO


class TDArray(Matrix):
    def __init__(self, dimensionality: int):
        super().__init__(dimensionality)
        self.matrix_data.elements = [[]]

    def input(self, read_file: IO) -> IO:
        size: int = self.matrix_data.dimensionality
        col: int = 0
        row: int = 0

        all_data = list(map(float, read_file.readline().split()))[::-1]
        self.matrix_data.elements = [
            [all_data.pop() for _ in range(size)]
            for _ in range(size)
        ]

        return read_file

    def input_random(self) -> None:
        size: int = self.matrix_data.dimensionality

        self.matrix_data.elements = [
            [get_rand_int() for _ in range(size)]
            for _ in range(size)
        ]

    def output(self, write_file: IO) -> None:
        size: int = self.matrix_data.dimensionality
        col: int = 0
        row: int = 0

        write_file.write(
            f'Two dimensional array:\n'
            f'Dimensionality: {size}\n'
            f'Arithmetic mean: {self.arithmetic_mean()}\n'
            f'Elements: \n'
        )

        for i in range(size ** 2):
            col = i % size
            row = i // size

            write_file.write(
                f'{self.matrix_data.elements[row][col]}'
                + (' ' if col != size - 1 else '\n')
            )

            col += 1

    def arithmetic_mean(self) -> float:
        size: int = self.matrix_data.dimensionality
        count: int = size ** 2
        elements_sum: float = 0
        col: int = 0
        row: int = 0

        for i in range(size ** 2):
            col = i % size
            row = i // size

            elements_sum += self.matrix_data.elements[row][col]

            col += 1

        return elements_sum / count
